### PR TITLE
feat(gas-tank): add gas tank

### PR DESCRIPTION
**Pull request by alvrs**
Moving the gas tank contract into the MUD monorepo

### CODE REVIEW

File: .changeset/red-carrots-sip.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"@latticexyz/gas-tank": patch
+---
+
+Adds the `GasTank` paymaster, which will allow MUD users to use smart account and share gas balance between apps.
```
Comment (holic) [inline .changeset/red-carrots-sip.md:5]:
```suggestion
Adds the `GasTank` paymaster, which will allow MUD users to use smart accounts and share gas balance between apps.
```

File: packages/gas-tank/src/PaymasterSystem.sol [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+      userAccount = _recoverCallWithSignature(userOp);
+      if (userAccount == address(0)) {
+        revert("Account is not registered as spender for any user");
+      }
+    }
+
+    // Require the user account to have sufficient balance
+    uint256 balance = UserBalances.get(userAccount);
+    if (maxCost > balance) {
+      revert("Insufficient user balance");
```
Comment (holic) [inline packages/gas-tank/src/PaymasterSystem.sol:73]: Should we have custom errors?

File: packages/gas-tank/worlds.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+  "17069": {
+    "address": "0xba0149de3486935d29b0e50dfcc9e61bd40ae095",
+    "blockNumber": 286789
+  },
+  "31337": {
+    "address": "0xc7b0b46dc0a1131adbdb8762a2773f2769abb527"
+  }
```
Comment (holic) [inline packages/gas-tank/worlds.json:8]: Should we gitignore this?

File: packages/gas-tank/src/PaymasterSystem.sol (hunk 2 above)
Comment (alvrs) [reply packages/gas-tank/src/PaymasterSystem.sol:73]: the reason i went with string errors here is that unfortunately these errors are not surfaced to users anyway (the entry point contract doesn't forward them) and string errors make it easier to debug what's going wrong in the tx trace if one doesn't have the ABI handy

File: packages/gas-tank/worlds.json (hunk 3 above)
Comment (alvrs) [reply packages/gas-tank/worlds.json:8]: the `31337` part is unnecessary but I think for the other chains it's actually useful to see where the gas tank is deployed to

**Comment by alvrs** [comment]
Merging this into #2653 because it's needed for the full local AA setup

**Pull request merged**